src/python/pants/bsp/util_rules/targets.py

```diff
@@ -64,23 +129,160 @@ class WorkspaceBuildTargetsHandlerMapping(BSPHandlerMapping):
+        Type[BSPBuildTargetsMetadataRequest]
+    ] = union_membership.get(BSPBuildTargetsMetadataRequest)
+    metadata_request_types_by_lang_id = {
+        metadata_request_type.language_id: metadata_request_type
+        for metadata_request_type in metadata_request_types
+    }
+    for tgt in targets:
+        for metadata_request_type in metadata_request_types:
+            field_set_type: Type[FieldSet] = metadata_request_type.field_set_type
+            if field_set_type.is_applicable(tgt):
+                field_sets_by_lang_id[metadata_request_type.language_id].add(
+                    field_set_type.create(tgt)
+                )
+                # lang_ids_by_field_set[field_set_type].add(metadata_request_type.language_id)
+
+    # TODO: Consider how to check whether the provided languages are compatible or whether compatible resolves
+    # selected.
+
+    # Request each language backend to provide metadata for the BuildTarget.
+    metadata_results = await MultiGet(
+        Get(
+            BSPBuildTargetsMetadataResult,
+            BSPBuildTargetsMetadataRequest,
+            metadata_request_types_by_lang_id[lang_id](field_sets=tuple(field_sets)),
+        )
+        for lang_id, field_sets in field_sets_by_lang_id.items()
+    )
+    metadata_results_by_lang_id = {
+        lang_id: metadata_result
+        for lang_id, metadata_result in zip(field_sets_by_lang_id.keys(), metadata_results)
+    }
+
+    # Pretend to merge the metadata into a single piece of metadata, but really just choose the metadata
+    # from the last provider.
+    metadata_merge_order = find_metadata_merge_order(
+        [metadata_request_types_by_lang_id[lang_id] for lang_id in field_sets_by_lang_id.keys()]
+    )
+    # TODO: None if no metadata obtained.
+    metadata = metadata_results_by_lang_id[metadata_merge_order[-1].language_id].metadata
+    digest = await Get(Digest, MergeDigests([r.digest for r in metadata_results]))
+
+    # Determine base directory for this build target.
+    # TODO: Use a source root?
+    targets_with_sources = [tgt for tgt in targets if tgt.has_field(SourcesField)]
+    sources_paths = await MultiGet(
+        Get(SourcesPaths, SourcesPathsRequest(tgt[SourcesField])) for tgt in targets_with_sources
+    )
+    _logger.info(f"sources_paths = {sources_paths}")
+    merged_source_files: set[str] = set()
+    for sp in sources_paths:
+        merged_source_files.update(sp.files)
+
+    source_roots_result = await Get(
+        SourceRootsResult, SourceRootsRequest, SourceRootsRequest.for_files(merged_source_files)
+    )
+    source_root_paths = {x.path for x in source_roots_result.path_to_root.values()}
+    if len(source_root_paths) == 0:
+        base_dir = build_root.pathlib_path
+    elif len(source_root_paths) == 1:
+        base_dir = build_root.pathlib_path.joinpath(list(source_root_paths)[0])
+    else:
+        raise ValueError("Multiple source roots not supported for BSP build target.")
```

Comment:
Probably helpful to say what the target is from the dict option, along with the source roots.

Reply:
I'm the only reader for this exception currently. Will improve messages once we decide on a final UX which won't happen until I get third-party indexing working.

Reply:
And also because we probably will support multiple source roots but that can be explored in a follow-on PR.